walletconnectv2/src/main/kotlin/org/walletconnect/walletconnectv2/engine/EngineInteractor.kt

```diff
@@ -136,67 +147,97 @@ class EngineInteractor {
         }
+        observePublishError(onResult)
+
+        relayRepository.subscribe(settledSession.topic)
+        relayRepository.publish(Topic(proposal.topic), encryptedMessage)
     }
 
-    fun reject(reason: String, topic: String) {
+    fun reject(reason: String, topic: String, onResult: (Result<Any>) -> Unit) {
         require(::relayRepository.isInitialized)
-        val sessionReject = PreSettlementSession.Reject(
-            id = generateId(),
-            params = Session.Failure(reason = reason)
-        )
-        val json: String = relayRepository.getSessionRejectionJson(sessionReject)
+        val sessionReject = PreSettlementSession.Reject(id = generateId(), params = Session.Failure(reason = reason))
+        val json: String = jsonRpcSerializer.trySerialize(PreSettlementSession.Reject::class.java, sessionReject)
         val (sharedKey, selfPublic) = crypto.getKeyAgreement(Topic(topic))
         val encryptedMessage: String = codec.encrypt(json, sharedKey, selfPublic)
+
+        observePublishAcknowledgement(onResult, topic)
+        observePublishError(onResult)
+
         relayRepository.publish(Topic(topic), encryptedMessage)
     }
 
-    fun disconnect(topic: String, reason: String) {
+    fun disconnect(topic: String, reason: String, onResult: (Result<Any>) -> Unit) {
         require(::relayRepository.isInitialized)
         val sessionDelete = PostSettlementSession.SessionDelete(id = generateId(), params = Session.DeleteParams(Reason(message = reason)))
-        val json = relayRepository.getSessionDeleteJson(sessionDelete)
+        val json = jsonRpcSerializer.trySerialize(PostSettlementSession.SessionDelete::class.java, sessionDelete)
         val (sharedKey, selfPublic) = crypto.getKeyAgreement(Topic(topic))
         val encryptedMessage: String = codec.encrypt(json, sharedKey, selfPublic)
-        //TODO Add subscriptionId from local storage
-        //TODO Delete all data from local storage coupled with given session
+
+        observePublishAcknowledgement(onResult, topic)
+        observePublishError(onResult)
+
+        //TODO Add subscriptionId from local storage + Delete all data from local storage coupled with given session
         relayRepository.unsubscribe(Topic(topic), SubscriptionId("1"))
         relayRepository.publish(Topic(topic), encryptedMessage)
     }
 
+    private fun observePublishError(onResult: (Result<String>) -> Unit) {
+        scope.launch {
+            relayRepository.observePublishResponseError
+                .catch { exception -> Timber.tag("WalletConnect exception").e(exception) }
+                .collect { errorResponse ->
+                    onResult(Result.failure(Throwable(errorResponse.error.errorMessage)))
+                    cancel()
+                }
+        }
+    }
+
+    private fun observePublishAcknowledgement(onResult: (Result<Any>) -> Unit, result: Any) {
```

Comment:
Change the function to accept `Result<T>`
```suggestion
    private fun <T> observePublishAcknowledgement(onResult: (Result<T>) -> Unit, result: T) {
```

Reply:
Good call